### PR TITLE
[BugFix] Fix the bug that causes incorrect pindex data when the key length is very long

### PR DESCRIPTION
**Pull request by sevev**
## Why I'm doing:
Pindex maybe return error result if the key is very long. We will calculate a hash value for each key, and use this hash value to determine which shard, page, and bucket the key will be distributed to.  And we will record the bucket info (include pageid, packid and size) in the `BucketHeader`. However, the `pageid` in BucketHeader is store as `uint16_t` and the max page num should be not greater than 65536. But now it's possible that the number of pages exceeds 65536. A typical example is following:
1. Assuming each key has a length of 3000 characters.
2. The page size is 4096 right now and each page can only keep one key right now.
3. We determine shard num by the total bytes of all the keys and we will try to generate a immutable pindex.
4. Since we use hashing to determine which page a key should be placed in, it's possible that different keys will be placed in the same page. If that, we will increase the page num and retry until each page only has on key at most. At this time, the page num maybe increase very fast and exceeds 65536.

## What I'm doing:
IMO, A 4k-sized page isn't suitable for storing very long keys, as this leaves little tolerance for hash collisions, resulting in many empty pages and causing performance degradation and storage space waste and When the key length is not very long, the likelihood of exceeding the limit of the number of pages is very small.  So, I supported larger pages to ensure that each page can store at least `kRecordPerBucket` (default value is 8) records to tolerate hash collisions. When the number of pages exceeds the limit, an error will be returned.

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by luohaha** [comment]
Should we support such long key length? It think it will be bad for performance even if we support it.

**Review by decster — DISMISSED** [review]

**Comment by luohaha** [comment]
If you want to backport this pr to 3.1, remember to set `PERSISTENT_INDEX_VERSION_5` when lake persistent index load from tablet.

**Review by luohaha — DISMISSED** [review]

File: gensrc/proto/persistent_index.proto [hunk 1]
```diff
@@ -48,6 +48,7 @@ message ImmutableIndexShardMetaPB {
     uint64 nbucket = 6;
     uint64 data_size = 7;
     uint64 uncompressed_size = 8; // if uncompressed_size is 0, which means no compression
+    uint64 page_size = 9; // if page_size is 0, which means upgrade from old version, we should initial it as 4096
```
[the remark this answers is not in the document]
Comment (sevev) [reply gensrc/proto/persistent_index.proto]: > 

Explicit default values are not allowed in proto3.

**Review by wyb — APPROVED** [review]

**Review by luohaha — APPROVED** [review]

**Comment by sevev** [comment]
@mergify backport branch-3.1

**Pull request merged**